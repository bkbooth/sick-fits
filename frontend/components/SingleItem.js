import React from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import Head from 'next/head';
import styled from 'styled-components';
import Error from 'components/ErrorMessage';

export const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM($itemId: ID!) {
    item(where: { id: $itemId }) {
      id
      title
      description
      largeImage
    }
  }
`;

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;

const SingleItem = ({ itemId }) => (
  <Query query={SINGLE_ITEM_QUERY} variables={{ itemId }}>
    {({ data: { item }, error, loading }) => {
      if (loading) return <p>Loading...</p>;
      if (error) return <Error error={error} />;
      if (!item) return <p>No Item found for ID '{itemId}'</p>;
      return (
        <SingleItemStyles>
          <Head>
            <title>{item.title} | Sick Fits!</title>
          </Head>
          <img src={item.largeImage} alt={item.title} />
          <div className="details">
            <h2>{item.title}</h2>
            <p>{item.description}</p>
          </div>
        </SingleItemStyles>
      );
    }}
  </Query>
);

export default SingleItem;
