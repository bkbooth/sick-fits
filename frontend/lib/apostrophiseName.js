const ESS = 's'

export default function apostrophiseName(name) {
  const nameEndsWithEss = name[name.length - 1].toLowerCase() === ESS
  return `${name}'${nameEndsWithEss ? '' : ESS}`
}
