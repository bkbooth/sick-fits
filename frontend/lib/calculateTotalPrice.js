/**
 * Calculate the total price of a cart by summing cart items' price multiplied by quantity
 *
 * @param {CartItem[]} cart
 * @param
 *
 * @returns {number}
 */
export default function calculateTotalPrice(cart) {
  return cart.reduce((total, cartItem) => {
    if (!cartItem.item) return total;
    return total + cartItem.quantity * cartItem.item.price;
  }, 0);
}

/**
 * @typedef {Object} CartItem
 * @property {number} quantity
 * @property {Object} item
 * @property {number} item.price
 */
