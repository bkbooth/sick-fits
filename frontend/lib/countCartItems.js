/**
 * Count the total number of items in a cart
 *
 * @param {CartItem[]} cart
 *
 * @returns {number}
 */
export default function countCartItems(cart) {
  return cart.reduce((total, cartItem) => total + cartItem.quantity, 0);
}

/**
 * @typedef {Object} CartItem
 * @property {number} quantity
 */
