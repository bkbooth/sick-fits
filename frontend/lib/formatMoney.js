export default function formatMoney(amount) {
  return new Intl.NumberFormat('en-AU', {
    style: 'currency',
    currency: 'AUD',
    // if its a whole, dollar amount, leave off the .00
    minimumFractionDigits: amount % 100 === 0 ? 0 : 2,
  }).format(amount / 100);
}
