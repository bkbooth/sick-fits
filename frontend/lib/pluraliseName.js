const ESS = 's'

export default function pluraliseName(name) {
  const nameEndsWithEss = name[name.length - 1].toLowerCase() === ESS
  return `${name}'${nameEndsWithEss ? '' : ESS}`
}
