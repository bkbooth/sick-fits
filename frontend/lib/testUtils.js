export const fakeItem = () => ({
  __typename: 'Item',
  id: 'abc123',
  title: 'Testing item',
  description: 'A really testable item',
  price: 5000,
  image: 'item.jpg',
  largeImage: 'yugeitem.jpg',
  user: null,
});
