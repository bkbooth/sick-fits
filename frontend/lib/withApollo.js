import ApolloClient, { InMemoryCache } from 'apollo-boost';
import withApollo from 'next-with-apollo';
import { LOCAL_STATE_QUERY } from 'components/Cart';
import { GRAPHQL_URL } from '../config';

function createApolloClient({ headers, initialState }) {
  return new ApolloClient({
    uri: GRAPHQL_URL,
    cache: new InMemoryCache().restore(initialState || {}),
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_parent, _args, { cache }, _info) {
            const { cartOpen } = cache.readQuery({ query: LOCAL_STATE_QUERY });
            const result = { data: { cartOpen: !cartOpen } };
            cache.writeData(result);
            return result;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo(createApolloClient);
