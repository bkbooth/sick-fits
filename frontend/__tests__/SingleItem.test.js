import { MockedProvider } from 'react-apollo/test-utils';
import { mount } from 'enzyme';
import wait from 'waait';
import { fakeItem } from 'lib/testUtils';
import SingleItem, { SINGLE_ITEM_QUERY } from 'components/SingleItem';

describe('<SingleItem />', () => {
  it('renders with proper data', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { itemId: '123' } },
        result: { data: { item: fakeItem() } },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem itemId="123" />
      </MockedProvider>
    );
    expect(wrapper.text()).toContain('Loading...');

    await wait();
    wrapper.update();
    expect(wrapper.find('h2')).toMatchSnapshot();
    expect(wrapper.find('img')).toMatchSnapshot();
    expect(wrapper.find('p')).toMatchSnapshot();
  });

  it('errors with a not found item', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { itemId: '123' } },
        result: { errors: [{ message: 'Item not found' }] },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem itemId="123" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('[data-test="graphql-error"]')).toMatchSnapshot();
  });
});
