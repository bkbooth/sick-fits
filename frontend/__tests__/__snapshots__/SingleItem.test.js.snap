// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> errors with a not found item 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Item not found
</p>
`;

exports[`<SingleItem /> renders with proper data 1`] = `
<h2>
  Testing item
</h2>
`;

exports[`<SingleItem /> renders with proper data 2`] = `
<img
  alt="Testing item"
  src="yugeitem.jpg"
/>
`;

exports[`<SingleItem /> renders with proper data 3`] = `
<p>
  A really testable item
</p>
`;
